pkg/engine/renderer/enhancer.go

```diff
@@ -83,6 +69,106 @@ func (k *DefaultEnhancer) Apply(templates map[string]string, metadata Metadata)
+	labelPaths := [][]string{
+		{"metadata", "labels"},
+		{"spec", "template", "metadata", "labels"},
+		{"spec", "volumeClaimTemplates[]", "metadata", "labels"},
+		{"spec", "jobTemplate", "metadata", "labels"},
+		{"spec", "jobTemplate", "spec", "template", "metadata", "labels"},
+	}
+
+	fieldsToAdd := map[string]string{
+		kudo.HeritageLabel: "kudo",
+		kudo.OperatorLabel: metadata.OperatorName,
+		kudo.InstanceLabel: metadata.InstanceName,
+	}
+
+	for _, path := range labelPaths {
+		if err := applyMapValues(obj, fieldsToAdd, path...); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func addAnnotations(obj map[string]interface{}, metadata Metadata) error {
+	// List of pathsfor annotations from here:
+	// https://github.com/kubernetes-sigs/kustomize/blob/master/api/konfig/builtinpluginconsts/commonannotations.go
+	annotationPaths := [][]string{
+		{"metadata", "annotations"},
+		{"spec", "template", "metadata", "annotations"},
+		{"spec", "jobTemplate", "metadata", "annotations"},
+		{"spec", "jobTemplate", "spec", "template", "metadata", "annotations"},
+	}
+
+	fieldsToAdd := map[string]string{
+		kudo.PlanAnnotation:            metadata.PlanName,
+		kudo.PhaseAnnotation:           metadata.PhaseName,
+		kudo.StepAnnotation:            metadata.StepName,
+		kudo.OperatorVersionAnnotation: metadata.OperatorVersion,
+		kudo.PlanUIDAnnotation:         string(metadata.PlanUID),
+	}
+
+	for _, path := range annotationPaths {
+		if err := applyMapValues(obj, fieldsToAdd, path...); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func applyMapValues(obj map[string]interface{}, fieldsToAdd map[string]string, path ...string) error {
+	for i, p := range path {
+		// If we have an element with a slice in the path, apply the fields to all elements of the
+		// slice with the remaining path
+		if strings.HasSuffix(p, "[]") {
+			sliceField := strings.TrimSuffix(p, "[]")
+
+			subPath := append(path[0:i], sliceField)
+			remainingPath := path[i+1:]
+
+			unstructuredSlice, found, err := unstructured.NestedSlice(obj, subPath...)
+			if !found || err != nil {
+				// We don't return err here, as it just means that path is invalid for this object.
+				// This is ok and does not indicate an error
+				return nil
+			}
+			for _, s := range unstructuredSlice {
+				if sliceMap, ok := s.(map[string]interface{}); ok {
+					if err = applyMapValues(sliceMap, fieldsToAdd, remainingPath...); err != nil {
+						return err
+					}
+				}
+			}
+			if err = unstructured.SetNestedSlice(obj, unstructuredSlice, subPath...); err != nil {
+				return err
+			}
+
+			return nil
+		}
+	}
+
+	// Merge added fields to map at specified path
+	stringMap, _, err := unstructured.NestedStringMap(obj, path...)
+	if err != nil {
+		// We don't return err here, as it just means that path is invalid for this object.
+		// This is ok and does not indicate an error
+		return nil
+	}
+	if stringMap == nil {
```

Comment:
this check seems not useful... `NestedStringMap` never returns a map == nil expect for err != nil which is captured above.

Reply:
It can return (nil, false, nil) if the path is valid but no map is in the structure. Maybe it would be more clear if we use found instead of relying on `stringMap == nil`?